KEYCLOAK-4414 Allow to add user-agent header for out-bound http calls

thomasdarimont opened the pull request:
We now support the configuration of `user-agent` headers set for out-bound http
calls performed by the Keycloak server to ease monitoring.

By default the user agent that is generated by the Apache HTTP Client is used, which currently exposes the underlying Java version, e.g.:
```
User-Agent: Apache-HttpClient/4.5.2 (Java/1.8.0_144)
```

With this users can configure the user-agent header via the SPI configuration to something less telling.
like `Keycloak (acme SSO)` or even just `acme SSO`.

jboss-cli example:
```
/subsystem=keycloak-server/spi=connectionsHttpClient/provider=default:write-attribute(name=properties.user-agent,value="Keycloak (acme SSO)")
```

thomasdarimont committed:
KEYCLOAK-4414 Add user-agent header for out-bound http calls

We now support the configuration of `user-agent` headers set for out-bound http
calls performed by the Keycloak server to ease monitoring.

By default the user agent that is generated by the Apache HTTP Client is used.

Users can configure the user-agent header via the SPI configuration.

jboss-cli example:
```
/subsystem=keycloak-server/spi=connectionsHttpClient/provider=default:write-attribute(name=properties.user-agent,value="Keycloak (acme SSO)")
```

stianst commented:
Would it make sense to just set the user-agent to Keycloak by default?

I'd like to see a test added for this.

stianst commented:
@thomasdarimont what do you think about setting the user-agent to Keycloak by default? Also, I'd like a test added for this.

thomasdarimont committed:
KEYCLOAK-4414 Revise DefaultHttpClientFactory to ease testing

Adds DefaultHttpClientFactoryTest with minimalistic infrastructure to test
HttpClient configuration.

thomasdarimont commented:
@stianst I think `Keycloak` as default user agent would be fine. I just rebased the code on top of current master and added 2 tests. 

btw. testing of keycloak components is often not easy due to the service / provider lookups.
It would be very helpful to design the components in such a way that they can be constructed 
with as little infrastructure as possible. This also helps to make the tests fast.

stianst commented:
We've made a decision to stick with functional/integration tests only using Arquillian and not to do any unit testing. So that's not needed at all. For the record we've discussed that several times in the lifespan of Keycloak and everytime come to the same conclusion.

The way you are testing it here is not acceptable to us. You need to test it in the Arquillian based testsuite and by making the server send proper requests.

thomasdarimont commented:
@stianst Thanks for your feedback I'll revise the PR with proper integration tests. This will take me a few hours which I'm willing to invest but I'd strongly suggest to rethink your testing approach. 
I known that your team discussed this several times already but I think there should be exceptions to the rules - sometimes an integration test is just overkill. I think the truth lies something in-between.

I'd love to discuss this over a beer or two in person one day :-)

stianst commented:
If you ask our QE team they won't accept unit tests. One especially nice thing about the tests we have at the moment is that you can run them against the embedded server or even a patched server (RH-SSO). That way you can check if patching the server actually fixes an issue. With unit tests we end up testing with the dependency versions specified in the poms and not by the runtime itself (remember Keycloak is on WildFly, while RH-SSO is EAP). I could go on ;)

We have the something in-between already, which can be used in some cases. That's the runOnServer thing:
https://github.com/keycloak/keycloak/blob/master/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/runonserver/RunOnServerTest.java

It's kinda white box integration testing approach. Wouldn't be suitable in this case though as it's primarily when you want to invoke a specific provider, but in this case you actually want to check things at the HTTP level.

stianst commented:
Oh and yes it would be great to have a beer. If you're even in Bergen, Norway, let me know ;)

stianst reviewed (changes requested):
Approach to testing is not following our standard approach and should be changed. Refactoring to DefaultHttpClientFactory to support unit testing should be removed. Instead a test should be written that mocks a client to receive a real HTTP request from Keycloak in order to check the actual HTTP headers sent. RealmTest.pushNotBefore can be used as inspiration. pushNotBefore sends an HTTP request to a test clients admin endpoint to push a new not before value.

stianst commented:
@thomasdarimont did you see my previous comment? Would be good to get this in, but I'd like to get rid of the refactoring and have a proper functional test for this.

stianst commented:
Closing due to lack of feedback. Feel free to re-open once comments have been addressed.

thomasdarimont commented:
Just removed the refactoring (locally) as you suggested, although I still think that it improved the code a bit...
If I understood you correctly, then you're suggesting that I could use `pushNotBefore` to trigger Keycloak to send a HTTP request to an endpoint controlled by the test (perhaps an embedded webserver... as in my previous testcase) to verify the user-agent header. Is that correct?

Where should I place the test?

stianst commented:
@thomasdarimont I'd just create a new HttpClientTest and base it on RealmTest.pushNotBefore

thomasdarimont commented:
Thanks for reopening.

I see that in `testrealm.json` there is a `test-app` client configured that is configured to accept kc_admin requests on the `http://localhost:8180/auth/realms/master/app/admin` endpoint.
```
    {
      "clientId": "test-app",
      "enabled": true,
      "baseUrl": "http://localhost:8180/auth/realms/master/app/auth",
      "redirectUris": [
        "http://localhost:8180/auth/realms/master/app/auth/*"
      ],
      "adminUrl": "http://localhost:8180/auth/realms/master/app/admin",
      "secret": "password"
    }
```
Is there already support for recording and later retrieving the HTTP request that get's sent to that client endpoint?

stianst commented:
Actually it's testingClient.testApp().getAdminPushNotBefore() that is getting the request that was received by the client. It doesn't currently have support to retrieve the actual request including headers, but that should be easy to add. Just add to TestApplicationResource (JAX-RS client interface) and implementation to TestApplicationResourceProvider. Remember whatever is returned needs to be serializable so probably not the HttpRequest, but rather a new object with I guess String entity and Map<String, String> headers.

stianst commented:
Closing due to lack of feedback. To get this accepted please send a new PR without the refactoring and with functional tests as suggested.
